ext/serializer.c

```diff
@@ -1787,111 +1792,134 @@ zval *ddtrace_serialize_span_to_array(ddtrace_span_data *span, zval *array) {
                 smart_str_appendc(&meta_str, '\'');
-            } ZEND_HASH_FOREACH_END();
+            }
             smart_str_0(&meta_str);
         }
+
+        if (meta_keys) {
+        }
+
         smart_str metrics_str = {0};
-        if (zend_hash_num_elements(Z_ARR(metrics_zv))) {
-            ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARR(metrics_zv), key, tag_zv) {
+        size_t metrics_count = 0;
+
+        ddog_CharSlice *metrics_keys = ddog_span_metrics_get_keys(rust_span, &metrics_count);
+        if (metrics_keys && metrics_count > 0) {
+            for (size_t i = 0; i < metrics_count; i++) {
+                ddog_CharSlice key = metrics_keys[i];
+                if (!key.ptr || key.len == 0) {
+                    continue;
+                }
+
+                double value;
+                bool found = ddog_get_span_metrics(rust_span, key, &value);
+                if (!found) {
+                    continue;
+                }
+
                 if (metrics_str.s) {
                     smart_str_appends(&metrics_str, ", ");
                 }
-                smart_str_append(&metrics_str, key);
+
+                smart_str_appends(&metrics_str, key.ptr);
                 smart_str_appends(&metrics_str, "='");
-                smart_str_append_double(&metrics_str, Z_DVAL_P(tag_zv), 12, false);
+                smart_str_append_double(&metrics_str, value, 12, false);
                 smart_str_appendc(&metrics_str, '\'');
-            } ZEND_HASH_FOREACH_END();
+            }
             smart_str_0(&metrics_str);
         }
-        prop_name = zend_hash_str_find(Z_ARR_P(el), ZEND_STRL("name")); // refetch, array may have been rehashed
+
+        ddog_CharSlice span_name = ddog_get_span_name(rust_span);
         log("Encoding span %" PRIu64 ": trace_id=%s, name='%s', service='%s', resource: '%s', type '%s' with tags: %s; and metrics: %s",
             span->span_id,
-            Z_STRVAL(span->root->property_trace_id), prop_name && Z_TYPE_P(prop_name) == IS_STRING ? Z_STRVAL_P(prop_name) : "",
+            Z_STRVAL(span->root->property_trace_id), span_name.ptr ? span_name.ptr : "",
             Z_TYPE(prop_service_as_string) == IS_STRING ? Z_STRVAL(prop_service_as_string) : "",
             Z_TYPE(prop_resource_as_string) == IS_STRING ? Z_STRVAL(prop_resource_as_string) : "",
             Z_TYPE(prop_type_as_string) == IS_STRING ? Z_STRVAL(prop_type_as_string) : "",
             meta_str.s ? ZSTR_VAL(meta_str.s) : "-",
             metrics_str.s ? ZSTR_VAL(metrics_str.s) : "-");
+
+        ddog_span_free_keys_ptr(meta_keys, meta_count);
+        ddog_span_free_keys_ptr(metrics_keys, metrics_count);
+
         smart_str_free(&meta_str);
         smart_str_free(&metrics_str);
-    })
-
-    if (zend_hash_num_elements(Z_ARR(metrics_zv))) {
-        zend_hash_str_add_new(Z_ARR_P(el), ZEND_STRL("metrics"), &metrics_zv);
-    } else {
-        zend_array_destroy(Z_ARR(metrics_zv));
-    }
+    });
```

Comment:
Suggestion; maybe we should provide an API function in data-pipeline which returns that ("span 123: trace_id=4567, ..."), so that we can reuse that code, and fully goes to what's actually in the rust span.